tools/codesigningtool/codesigningtool.py

```diff
@@ -120,16 +123,100 @@ def _certificate_fingerprint(identity):
   fingerprint = fingerprint.replace(":", "")
   return fingerprint
 
+def _certificate_common_name(cert):
+  _, subject, _ = execute.execute_and_filter_output([
+    "openssl",
+    "x509",
+    "-noout",
+    "-inform",
+    "DER",
+    "-subject"
+  ], inputstr=cert, raise_on_failure=True)
+  subject = subject.strip().split('/')
+  cert_cn = [f for f in subject if "CN=" in f][0]
+  cert_cn = cert_cn.replace("CN=", "")
+  return cert_cn
 
 def _get_identities_from_provisioning_profile(mpf):
   """Iterates through all the identities in a provisioning profile, lazily."""
   for identity in mpf["DeveloperCertificates"]:
-    if not isinstance(identity, bytes):
+    cert = _certificate_data(identity)
+    yield _certificate_fingerprint(cert)
+
+def _certificate_data(cert):
+  if not isinstance(cert, bytes):
       # Old versions of plistlib return the deprecated plistlib.Data type
       # instead of bytes.
-      identity = identity.data
-    yield _certificate_fingerprint(identity)
+      cert = cert.data
+
+  return cert
 
+def _find_smart_card_identities(identity=None):
+  """Finds smartcard identitites on the current system."""
+  ids = []
+  # Use 'system_profiler' to export XML
+  # containing smartcards keychain data
+  _, xml, _ = execute.execute_and_filter_output([
+      "/usr/sbin/system_profiler",
+      "SPSmartCardsDataType",
+      "-xml"
+  ], raise_on_failure=True)
+  xml = ET.fromstring(xml)
+  # Find 'AVAIL_SMARTCARDS_TOKEN' node and move up one level
+  # to find the 'array' of available tokens, each one is a 'string' node
+  available_tokens_xml = xml.findall('.//*/string[.="AVAIL_SMARTCARDS_TOKEN"]../array')[0]
```

Comment:
We may want to not assume that these `AVAIL_SMARTCARDS_TOKEN` and `AVAIL_SMARTCARDS_KEYCHAIN` (below) items exist, and default back to an empty array that could help the function return earlier

Reply:
That makes sense, will add the fallback logic

Reply:
Using `plistlib` now and getting these keys in a way that empty arrays are returned if not present